compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -4613,67 +4536,14 @@ TR::Register *OMR::Power::TreeEvaluator::ZEROCHKEvaluator(TR::Node *node, TR::Co
-      {
-      if (valueToCheck->getOpCode().isCompareForOrder())
-         {
-         if (valueToCheck->getChild(0)->getOpCode().is8Byte() ||
-             valueToCheck->getChild(1)->getOpCode().is8Byte())
-            {
-            // switch branches since we want to go to OOL on node evaluating to 0
-            // which corresponds to an ifcmp fall-through
-            TR::InstOpCode::Mnemonic branchOp = cmp2branch(valueToCheck->getOpCode().getOpCodeForReverseBranch(), cg);
-            TR::InstOpCode::Mnemonic reverseBranchOp = cmp2branch(valueToCheck->getOpCodeValue(), cg);
-            TR::TreeEvaluator::compareLongsForOrder(branchOp, reverseBranchOp, slowPathLabel, valueToCheck, cg,
-                                 valueToCheck->getOpCode().isUnsignedCompare(), true, PPCOpProp_BranchUnlikely);
-            }
-         else
-            {
-            // switch branches since we want to go to OOL on node evaluating to 0
-            // which corresponds to an ifcmp fall-through
-            TR::TreeEvaluator::compareIntsForOrder(cmp2branch(valueToCheck->getOpCode().getOpCodeForReverseBranch(), cg),
-                                slowPathLabel, valueToCheck, cg, valueToCheck->getOpCode().isUnsignedCompare(),
-                                true, PPCOpProp_BranchUnlikely);
-            }
-         }
-      else
-         {
-         TR_ASSERT(valueToCheck->getOpCode().isCompareForEquality(), "Compare opcode must either be compare for order or for equality");
-         if (valueToCheck->getChild(0)->getOpCode().is8Byte() ||
-             valueToCheck->getChild(1)->getOpCode().is8Byte())
-            {
-            // switch branches since we want to go to OOL on node evaluating to 0
-            // which corresponds to an ifcmp fall-through
-            TR::TreeEvaluator::compareLongsForEquality(cmp2branch(valueToCheck->getOpCode().getOpCodeForReverseBranch(), cg),
-                                    slowPathLabel, valueToCheck, cg, true, PPCOpProp_BranchUnlikely);
-            }
-         else
-            {
-            TR::TreeEvaluator::compareIntsForEquality(cmp2branch(valueToCheck->getOpCode().getOpCodeForReverseBranch(), cg),
-                                   slowPathLabel, valueToCheck, cg, true, PPCOpProp_BranchUnlikely);
-            }
-         }
-      }
-   else
-      {
-      TR::Register *value = cg->evaluate(node->getFirstChild());
-      TR::Register *condReg = cg->allocateRegister(TR_CCR);
-      if (cg->comp()->target().is64Bit())
-         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi8, node, condReg, value, 0);
-      else
-         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, value, 0);
-      generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, PPCOpProp_BranchUnlikely, node, slowPathLabel, condReg);
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
+   CompareCondition cond = evaluateToConditionRegister(condReg, node, node->getFirstChild(), cg);
```

Comment:
This is for the case where the first child node is not a compare node.

The implementation of `evaluateToConditionRegister` appears to assume that the third parameter is a node that will evaluate to an 4 byte value. It's not clear to me why this assumption can be made.

https://github.com/eclipse/omr/blob/a8cb0173e164d857350bca95fff6fce3aa92b807/compiler/p/codegen/ControlFlowEvaluator.cpp#L1060-L1062

The original `ZEROCHKEvaluator` handles the case where `node->getFirstChild()` can evaluate to an 8 byte value.

Reply:
I'm not entirely sure why the previous implementation of `ZEROCHK` on Power had code for handling this, since the IL opcode properties of `ZEROCHK` make it clear that its first child must be an `Int32`:

https://github.com/eclipse/omr/blob/a8cb0173e164d857350bca95fff6fce3aa92b807/compiler/il/OMRILOpCodeProperties.hpp#L5681

It's not possible to have an `Int64`-typed node as the first child of a `ZEROCHK` node in well-formed trees, as any such use would be considered malformed by the IL validator.

Reply:
Okay, that makes sense to me.

Reply:
Interesting, in fact,  it seems like we always used cmpi8 in 64-bit mode... Maybe it was a hidden bug?